File: typescript/009_dnd-character/dnd-character.ts
export class DnDCharacter {
  public strength: number;
  public dexterity: number;
  public constitution: number;
  public intelligence: number;
  public wisdom: number;
  public charisma: number;
  public hitpoints: number;

  constructor() {
    this.strength = DnDCharacter.generateAbilityScore();
    this.dexterity = DnDCharacter.generateAbilityScore();
    this.constitution = DnDCharacter.generateAbilityScore();
    this.intelligence = DnDCharacter.generateAbilityScore();
    this.wisdom = DnDCharacter.generateAbilityScore();
    this.charisma = DnDCharacter.generateAbilityScore();
    this.hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution);
  }
  
  private static getRandomInt = (min: number, max: number): number => 
    Math.floor(Math.random() * (max - min + 1)) + min;
    
  public static generateAbilityScore(): number {
    const dices: number[] = Array.from({ length: 4 }, () => DnDCharacter.getRandomInt(1, 6));
    return dices.reduce((acc, i) => acc + i) - Math.min(...dices);
  }

  public static getModifierFor = (abilityValue: number): number => 
    Math.floor((abilityValue - 10) / 2);
}

File: python/ex_070_dnd-character/dnd_character.py
import random

ABILITIES = ('strength', 'dexterity', 'constitution', 'intelligence', 'wisdom', 'charisma')

class Character:
    def __init__(self):
        for stat_name in ABILITIES:
            setattr(self, stat_name, self.ability())
        self.hitpoints = 10 + modifier(self.constitution)

    def ability(self):
        four_dice_throws = [random.randint(1, 6) for _ in range(4)]
        return sum(four_dice_throws) - min(four_dice_throws)

def modifier(value):
    return (value - 10) // 2

    
